### PR TITLE
Add net.hachune.mikutter

### PR DESCRIPTION
**Pull request by yuntan**
mikutter is a twitter client. https://mikutter.hachune.net/

### CODE REVIEW

File: net.hachune.mikutter.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    <screenshot>
+      <image width="427" height="284" xml:lang="ja">https://mikutter.hachune.net/img/mikutter-multi-account.png</image>
+      <caption>Multi account</caption>
+      <caption xml:lang="ja">マルチアカウント</caption>
+    </screenshot>
+  </screenshots>
+  ⁠<update_contact>yuntan.sub1_AT_gmail.com</update_contact>
+  <translation type="gettext">mikutter</translation>
+  <content_rating type="oars-1.1">
+    <content_attribute id="violence-cartoon">none</content_attribute>
```
Comment (x80486) [inline net.hachune.mikutter.appdata.xml]: If you use `oars-1.1` you can remove all the `content_attribute` tags with `none` value :sunglasses:

Comment (yuntan) [reply net.hachune.mikutter.appdata.xml]: Thanks 🙂

**Comment by nedrichards** [comment]
bot, build net.hachune.mikutter

**Comment by nedrichards** [comment]
@yuntan without being rude about it, given twitter's new API changes, will this application continue to work?

**Comment by yuntan** [comment]
mikutter supports Twitter's API 1.1 endpoint so it is functional except for streaming after UserStream API closing. mikutter's consumer key/secret is suspended in May 2018 and users need to obtain CK/CS by themselves to use Twitter functionally. I'll add the instruction to appdata.xml later.

In addition, [thirdparty mastodon plugin](https://github.com/cobodo/mikutter-worldon) is available and this app is in demand among mastodon users.

File: flatpak_rubygems_generator.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
```
Comment (barthalion) [inline flatpak_rubygems_generator.rb]: Is it any different from https://github.com/flatpak/flatpak-builder-tools/blob/master/rubygems/flatpak_rubygems_generator.rb ? If not, just remove it or add as git submodule for convenience.

File: net.hachune.mikutter.appdata.xml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (barthalion) [inline net.hachune.mikutter.appdata.xml]: It should be maintained in upstream repo so other distributions can use it too.

File: net.hachune.mikutter.git-source.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (barthalion) [inline net.hachune.mikutter.git-source.json]: Is it really worth it to keep it separately?

File: net.hachune.mikutter.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  # X11 + XShm access
+- --share=ipc
+- --socket=x11
+  # Wayland access
+- --socket=wayland
+- --socket=pulseaudio
+  # Notification access
+- --talk-name=org.freedesktop.Notifications
+- --share=network
+- --filesystem=home
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Wouldn't xdg-pictures be good enough?

File: net.hachune.mikutter.yaml [hunk 6]
```diff
@@ -0,0 +1,69 @@
+app-id: net.hachune.mikutter
+runtime: org.gnome.Platform
+runtime-version: 3.28
+sdk: org.gnome.Sdk
+command: mikutter
+separate-locales: false
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Why is that, it doesn't provide locales?

File: net.hachune.mikutter.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  sources:
+  - type: archive
+    url: 'http://pyyaml.org/download/libyaml/yaml-0.1.7.tar.gz'
+    sha256: 8088e457264a98ba451a90b8661fcb4f9d6f478f7265d48322a196cec2480729
+- name: ruby
+  config-opts:
+  - --disable-install-doc
+  sources:
+  - type: archive
+    url: 'https://cache.ruby-lang.org/pub/ruby/2.4/ruby-2.4.4.tar.gz'
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Anything wrong with ruby 2.5?

File: net.hachune.mikutter.yaml [hunk 8]
```diff
@@ -0,0 +1,69 @@
+app-id: net.hachune.mikutter
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Please reformat it according to our [style guide](https://github.com/flathub/flathub/wiki/YAML-Style-Guide).

**Comment by nedrichards** [comment]
@yuntan do you plan to provide support for the mastadon plugin as an extension?

**Comment by yuntan** [comment]
@barthalion Yes, the upstream team will continue development of mikutter as a multi-service client. They are also interested in the distribution of flatpak and cooperative with me. Upstream discussion thread is [here](https://dev.mikutter.hachune.net/issues/1248).

@nedrichards No. One of the most advantages of mikutter is "anyone can write plugins, anyone can use it". Plugins which are officially judged as convenient for every mikutter user are included to mikutter itself but others are distributed by plugin author.

**Comment by nedrichards** [comment]
@yuntan right, but if there isn't an extension point defined in this flatpak for other plugins to be applied, then "noone can write plugins and noone can use them" because everything will be read only once you ship the flatpak.

**Comment by yuntan** [comment]
I’m sorry for the delay in my response. I'm going to be inactive for next 1 week. I'll fix and reply in the week after next. Thank you for your cooperation.

**Comment by nedrichards** [comment]
OK, that's fine - we understand that people have things to do, I appreciate your comment though.

File: net.hachune.mikutter.git-source.json (hunk 4 above)
Comment (yuntan) [reply net.hachune.mikutter.git-source.json]: Yes. This file is imported by two files (net.hachune.mikutter.yaml and net.hachune.mikutter.rubygems-module.json).

File: net.hachune.mikutter.yaml (hunk 5 above)
Comment (yuntan) [reply net.hachune.mikutter.yaml]: mikutter need to access ~/.mikutter and this line should be `--filesystem=~/.mikutter`, however the gtk theme and font setting is not applied without `--filesystem=home`. Is it the flatpak spec or a bug?

**Comment by yuntan** [comment]
@nedrichards mikutter can load plugins from ~/.mikutter/plugin so users can use their favorite plugin with flatpak mikutter.

File: net.hachune.mikutter.yaml (hunk 5 above)
Comment (barthalion) [reply net.hachune.mikutter.yaml]: No, it should be `--persit=.mikutter` or to be perfectly correct, start following XDG specs and use files from `$XDG_CONFIG_DIR/.mikutter` (which is `~/.config` by default). If it's GTK3 app, theme should be packaged separately and put on Flathub.

File: net.hachune.mikutter.git-source.json (hunk 4 above)
Comment (barthalion) [reply net.hachune.mikutter.git-source.json]: Ah, okay, I didn't notice.

File: net.hachune.mikutter.yaml (hunk 5 above)
Comment (yuntan) [reply net.hachune.mikutter.yaml]: Thank you for your advice.

> start following XDG specs

migration to path following XDG specs is under discussion.

>  the gtk theme and font setting is not applied without `--filesystem=home`

It was just my environment problem (installing theme package and switching to wayland session solved it).

File: net.hachune.mikutter.appdata.xml (hunk 3 above)
Comment (yuntan) [reply net.hachune.mikutter.appdata.xml]: I'll contact upstream to include appdata file.

**Comment by nedrichards** [comment]
bot, build net.hachune.mikutter

File: net.hachune.mikutter.yaml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+runtime: org.gnome.Platform
+runtime-version: 3.28
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Would be nice to test 3.30. Also it should be quoted to be a string, not a float.

File: net.hachune.mikutter.yaml [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  # X11 + XShm access
+  - --share=ipc
+  - --socket=x11
+  # Wayland access
+  - --socket=wayland
+  - --socket=pulseaudio
+  # Notification access
+  - --talk-name=org.freedesktop.Notifications
+  - --share=network
+  - --persist=.mikutter
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: Ideally it should use $XDG_CONFIG_DIR (something to report to upstream).

File: net.hachune.mikutter.yaml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+build-options:
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: These can be removed, runtimes based on 18.08 use their own hardened flags.

File: net.hachune.mikutter.yaml [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      - type: file
+        url: 'https://rubygems.org/downloads/bundler-1.16.2.gem'
+        sha256: 3bb53e03db0a8008161eb4c816ccd317120d3c415ba6fee6f90bbc7f7eec8690
+
+  - net.hachune.mikutter.rubygems-module.json
+
+  - name: mikutter
+    buildsystem: simple
+    build-commands:
+      - install -Dm644 -t /app/share/mikutter mikutter.rb LICENSE README
```
Comment (barthalion) [inline net.hachune.mikutter.yaml]: No need to install LICENSE and README unless they are displayed in-app; all files are accessible as part of `${app_id}.Source` extension.

**Comment by nedrichards** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
Sorry for silence. I'm really busy now (and couple month ahead) and cannot
get time to touch. Thanks for caring.

2019年3月1日(金) 5:38 Flathub Bot <notifications@github.com>:

File: net.hachune.mikutter.yaml (hunk 10 above)
Comment (yuntan) [reply net.hachune.mikutter.yaml]: Using `$XDG_CONFIG_DIR` is disscussed in upstream, but it seems difficult task (migration of existing configuration, separate cache between multiple instances, etc.). TBD

Upstream ticket: https://dev.mikutter.hachune.net/issues/1189

File: net.hachune.mikutter.yaml (hunk 9 above)
Comment (yuntan) [reply net.hachune.mikutter.yaml]: Updating to 3.30 brakes build. The native extension of gtk2 Ruby Gem cannot be built.

```
Installing gtk2 3.2.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /app/lib/ruby/gems/2.5.0/gems/gtk2-3.2.1/ext/gtk2
/app/bin/ruby -r ./siteconf20190503-3-1tbni39.rb extconf.rb
checking for --enable-debug-build option... no
checking for -Wall option to compiler... yes
checking for -Waggregate-return option to compiler... yes
checking for -Wcast-align option to compiler... yes
checking for -Wextra option to compiler... yes
checking for -Wformat=2 option to compiler... yes
checking for -Winit-self option to compiler... yes
checking for -Wlarger-than-65500 option to compiler... yes
checking for -Wmissing-declarations option to compiler... yes
checking for -Wmissing-format-attribute option to compiler... yes
checking for -Wmissing-include-dirs option to compiler... yes
checking for -Wmissing-noreturn option to compiler... yes
checking for -Wmissing-prototypes option to compiler... yes
checking for -Wnested-externs option to compiler... yes
checking for -Wold-style-definition option to compiler... yes
checking for -Wpacked option to compiler... yes
checking for -Wp,-D_FORTIFY_SOURCE=2 option to compiler... yes
checking for -Wpointer-arith option to compiler... yes
checking for -Wswitch-default option to compiler... yes
checking for -Wswitch-enum option to compiler... yes
checking for -Wundef option to compiler... yes
checking for -Wout-of-line-declaration option to compiler... no
checking for -Wunsafe-loop-optimizations option to compiler... yes
checking for -Wwrite-strings option to compiler... yes
checking for rb_errinfo()... yes
checking for cairo... yes
checking for rb_cairo.h... yes
checking for Windows... no
checking for gtk+-2.0 version (>= 2.10.0)... no
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.
```

**Comment by LeandroStanger** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
I forgot about this PR until yesterday... Which things are left to do?

**Comment by nedrichards** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
Upstream URL has chenged. See https://mikutter.hachune.net/develop .

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by yuntan** [comment]
bot, build net.hachune.mikutter

**Comment by nedrichards** [comment]
bot, build net.hachune.mikutter

**Pull request closed without merging**